Manifold/Constraint.swift

```diff
@@ -32,3 +59,78 @@ public func == (left: Constraint, right: Constraint) -> Bool {
 public func === (left: Type, right: Type) -> Constraint {
 	return Constraint(equality: left, right)
 }
+
+
+// MARK: - ConstraintSet
+
+public typealias ConstraintSet = Multiset<Constraint>
+
+private func typeGraph(constraints: ConstraintSet) -> (DisjointSet<Type>, [Type: Int]) {
+	let distinctTypes = Set(lazy(constraints)
+		.flatMap {
+			$0.analysis(ifEquality: {
+				$0.instantiate().distinctTypes.union($1.instantiate().distinctTypes)
+			})
+		})
+	let typeGraph = DisjointSet(distinctTypes)
+	return (typeGraph, Dictionary(lazy(enumerate(typeGraph)).map { ($1, $0) }))
+}
+
+private func structural<T>(t1: Type, t2: Type, initial: T, f: (T, Type, Type) -> T) -> T {
+	return
+		(t1.function &&& t2.function).map {
+			structural($0.0, $1.0, structural($0.1, $1.1, f(initial, t1, t2), f), f)
+		}
+	??	f(initial, t1, t2)
+}
+
+
+public func occurs(v: Variable, t: Type) -> Bool {
```

Comment:
In [_Generalizing Hindley-Milner Type Inference Algorithms_](http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.18.9348) §6, _Unbiased Constraint Solving_, this is instead represented by constructing term and type graphs; detection of infinite types amounts to a directed edge in the term graph connecting two elements within the same equivalence class in the type graph. For example, given this constraint:

```
τ ≡ τ -> Bool
```

the term graph looks like this:

```
  ->
 ↙︎₀ ↘︎₁
τ   Bool
```

and we have these equivalence classes:

- `{τ, (τ -> Bool)}`
- `{Bool}`

leading to a type graph like this:

```
  ->
⤢↙︎₀ ↘︎₁
τ   Bool
```

The existence of both `↙︎` and `⤢` edges between two types indicates a cyclic, and therefore infinite, type.

We instead use the unsophisticated approach above due to the (current) inconvenience of representing the term graph. I believe that extracting & employing [TesseractCore](https://github.com/antitypical/TesseractCore)’s [`Graph`](https://github.com/antitypical/TesseractCore/blob/master/TesseractCore/Graph.swift) will help considerably, as it will enable us to replace the construction of the `Set` of free variables and the (admittedly O(1)) membership test with O(α(n)) (where α(n) is the inverse of the [Ackermann function](http://en.wikipedia.org/wiki/Ackermann_function)) `find` in the already-constructed `DisjointSet` representing the type graph.